### PR TITLE
feat: add ability to delete by filtering metadata only

### PR DESCRIPTION
**Pull request by jeanmaried**
## What kind of change does this PR introduce?

Feature

## What is the current behavior?

I need to be able to delete vectors based solely on their metadata but currently the only way to do so is to already have the ids.

## What is the new behavior?

I added 2 methods as suggestions.
Option #1: add a `delete_by_metadata`
Option #2: `delete_vectors` (temporary name) a suggestion for updating the current `delete` function to allow optionally passing the ids or metdata (one or the other).

Please let me know which is an acceptable approach and if there are any objections to having this capability.

### CODE REVIEW

**Comment by olirice** [comment]
Deleting by metadata sounds reasonable. I think the right way to approach it would be to overload the existing `delete` method in a backwards compatible way

Something like:

From
```python
def delete(self, ids: Iterable[str]) -> List[str]:
```
To
```python
def delete(self, ids: Optional[Iterable[str]] = None, filters: Optional[Metadata] = None) -> List[str]:
```

where `filters` goes through the same `build_filter` logic as `query`.

**Comment by jeanmaried** [comment]
@olirice thank you, I updated the method and the docs.

File: src/vecs/collection.py [hunk 1]
```diff
@@ -340,33 +340,46 @@ def fetch(self, ids: Iterable[str]) -> List[Record]:
...
             ids (Iterable[str]): An iterable of vector identifiers.
+            filters (Metadata): A dictionary of metadata key-value pairs to match.
 
         Returns:
             List[str]: A list of the identifiers of the deleted vectors.
         """
-        if isinstance(ids, str):
-            raise ArgError("ids must be a list of strings")
+        if ids is None and filters is None:
+            raise VectorDeletionError("Either ids or metadata must be provided.")
```
Comment (olirice) [inline src/vecs/collection.py:355]: `VectorDeletionError` is not defined. Did you intended to add this in `exc.py`?
if so, it should inherit from ArgError to maintain compatibility

File: src/vecs/collection.py [hunk 2]
```diff
@@ -340,33 +340,46 @@ def fetch(self, ids: Iterable[str]) -> List[Record]:
...
 
-        chunk_size = 12
+        del_ids = []
 
-        del_ids = list(ids)
-        ids = []
         with self.client.Session() as sess:
             with sess.begin():
-                for id_chunk in flu(del_ids).chunk(chunk_size):
+                if ids is not None:
```
Comment (olirice) [inline src/vecs/collection.py:361]: lets move all client side checks outside the database transaction to keep them as fast as possible.
coalescing `None` to an empty list would probably work best for keeping the logic in the db transaction clear

File: src/vecs/collection.py [hunk 3]
```diff
@@ -340,33 +340,46 @@ def fetch(self, ids: Iterable[str]) -> List[Record]:
...
+                        raise ArgError("ids must be a list of strings")
+                    for id_chunk in flu(ids).chunk(12):
+                        stmt = (
+                            delete(self.table)
+                            .where(self.table.c.id.in_(id_chunk))
+                            .returning(self.table.c.id)
+                        )
+                        del_ids.extend(sess.execute(stmt).scalars() or [])
+                
+                if filters is not None:
```
Comment (olirice) [inline src/vecs/collection.py:372]: same here

**Review by olirice — CHANGES_REQUESTED** [review]
This PR introduces some LOC that aren't covered by the test suite. Could you add tests to handle the new uses case? 

These are the uncovered lines:
Screenshot 2023-11-06 at 9 52 02 AM


Screenshot 2023-11-06 at 9 52 02 AM

File: src/tests/test_collection.py [hunk 4]
```diff
@@ -101,14 +101,33 @@ def test_delete(client: vecs.Client) -> None:
...
     # bad input
     with pytest.raises(vecs.exc.ArgError):
         movies.delete(ids="should_be_a_list")
 
+    # bad input: neither ids nor filters provided.
+    with pytest.raises(vecs.exc.ArgError):
+        movies.delete()
+
+    # bad input: should only provide either ids or filters, not both
+    with pytest.raises(vecs.exc.ArgError):
```
Comment (jeanmaried) [inline src/tests/test_collection.py:127]: @olirice I'm a bit stumped on why this test fails. Any ideas? If both ids and filters are not None, I raise an error in the delete function. Any idea if I am missing something here?

Comment (jeanmaried) [reply src/tests/test_collection.py:127]: Seems to be passing now. Must have been a cache issue.

**Review by olirice — APPROVED** [review]

**Comment by olirice** [comment]
very nice, thank you!

**Pull request merged**